api/src/main/java/com/netflix/iceberg/io/FileIO.java

```diff
@@ -47,4 +46,30 @@
    * Delete the file at the given path.
    */
   void deleteFile(String path);
+
+  /**
+   * Get an {@link InputFile} to get the bytes for this table's metadata file with the given name.
+   */
+  InputFile readMetadataFile(String fileName);
+
+  /**
+   * Get an {@link OutputFile} to write bytes for a new table metadata file with the given name.
+   */
+  OutputFile newMetadataFile(String fileName);
+
+  /**
+   * Get an {@link OutputFile} for writing bytes to a new data file for this table.
+   * <p>
+   * The partition values of the rows in this file may be used to derive the final location of
+   * the file.
+   */
+  OutputFile newPartitionedDataFile(
+      PartitionSpec partitionSpec, StructLike partitionData, String fileName);
```

Comment:
It seems strange to me to pass in the partition spec and partition data. Why not have a single method, `newDataOutputFile`, that accepts a relative path?

Reply:
Passing in specifically the partitioning information here I think carries a stronger notion that this will be used for data with those partitionings, as opposed to a path in which callers can really pass in any string. So the stronger typing carries stronger semantics. It also signifies that the partitioning need not be encoded in the path, but the implementation can choose to derive its own paths based on the partitioning.

Reply:
> partitioning need not be encoded in the path

Good point. What about naming both `newDataOutputFile`? The arguments still carry the notion that this is for the partitioned case.

Reply:
Do we need the second unpartitioned variant? For example, what if we passed `PartitionSpec#unpartitioned` and an empty `PartitionData` structure instead? Or do we think that the API should make unpartitioned files a first-class concept?

Reply:
I think it is easier for implementers to have two methods. We could add a default implementation that calls the partitioned version and let implementations choose, but I think it is much easier to implement the unpartitioned case without worrying about the spec.